components/locid/src/helpers.rs

```diff
@@ -261,9 +402,150 @@ macro_rules! impl_writeable_for_single_subtag {
+                use $crate::$($full_name)::+;
+                const R: $name =
+                    match $name::from_bytes($string.as_bytes()) {
+                        Ok(r) => r,
+                        #[allow(clippy::panic)] // const context
+                        _ => panic!(concat!("Invalid ", stringify!($name), ": ", $string)),
+                    };
+                R
+            }};
+        }
+
+        #[cfg(feature = "databake")]
+        impl databake::Bake for $name {
+            fn bake(&self, env: &databake::CrateEnv) -> databake::TokenStream {
+                env.insert("icu_locid");
+                let string = self.as_str();
+                databake::quote! {::icu_locid::$macro_name!(#string) }
+            }
+        }
+
+        #[test]
+        fn test_construction() {
+            let maybe = $name::from_bytes($good_example.as_bytes());
+            assert!(maybe.is_ok());
+            assert_eq!(maybe, $name::try_from_raw(maybe.unwrap().into_raw()));
+            assert_eq!(maybe.unwrap().as_str(), $good_example);
+            $(
+                let maybe = $name::from_bytes($more_good_examples.as_bytes());
+                assert!(maybe.is_ok());
+                assert_eq!(maybe, $name::try_from_raw(maybe.unwrap().into_raw()));
+                assert_eq!(maybe.unwrap().as_str(), $more_good_examples);
+            )*
+            assert!($name::from_bytes($bad_example.as_bytes()).is_err());
+            $(
+                assert!($name::from_bytes($more_bad_examples.as_bytes()).is_err());
+            )*
+        }
+
         #[test]
         fn test_writeable() {
-            writeable::assert_writeable_eq!(&$type::from_str($sample).unwrap(), $sample);
+            writeable::assert_writeable_eq!(&$good_example.parse::<$name>().unwrap(), $good_example);
+            $(
+                writeable::assert_writeable_eq!($more_good_examples.parse::<$name>().unwrap(), $more_good_examples);
+            )*
+        }
+
+        #[cfg(feature = "serde")]
+        impl<'de> serde::Deserialize<'de> for $name {
+            fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+            where
+                D: serde::de::Deserializer<'de>,
+            {
+                struct Visitor;
+
+                impl<'de> serde::de::Visitor<'de> for Visitor {
+                    type Value = $name;
+
+                    fn expecting(
+                        &self,
+                        formatter: &mut core::fmt::Formatter<'_>,
+                    ) -> core::fmt::Result {
+                        write!(formatter, "a valid BCP-47 {}", stringify!($name))
+                    }
+
+                    fn visit_str<E: serde::de::Error>(self, s: &str) -> Result<Self::Value, E> {
+                        s.parse().map_err(serde::de::Error::custom)
+                    }
+                }
+
+                if deserializer.is_human_readable() {
+                    deserializer.deserialize_string(Visitor)
```

Comment:
It will give different error messages though. This uses "expected a valid BCP-47 language", whereas deserializing to string and then parsing would throw two different errors.

Reply:
(You could still get "expected a valid BCP-47 language", but only after the string passed UTF-8 validation)

Reply:
Yes, and that's annoying because you only reveal what you actually want piece by piece